source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * Trigger state machine action base on the state.
+ *
+ **/
+static void prvTriggerAction( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    switch( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState )
+    {
+        case azureiotprovisioningWF_STATE_CONNECT:
+            prvAzureIoTProvisioningClientConnect( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_SUBSCRIBE:
+            prvAzureIoTProvisioningClientSubscribe( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_REQUEST:
+            prvAzureIoTProvisioningClientRequest( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_RESPONSE:
+            prvAzureIoTProvisioningClientParseResponse( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_SUBSCRIBING:
+        case azureiotprovisioningWF_STATE_REQUESTING:
+        case azureiotprovisioningWF_STATE_COMPLETE:
+            /* None action taken here, as these states are waiting for receive path. */
+            break;
+
+        case azureiotprovisioningWF_STATE_WAITING:
+            prvAzureIoTProvisioningClientCheckTimeout( xAzureIoTProvisioningClientHandle );
+            break;
+
+        default:
+            AZLogError( ( "State not handled %d\r\n",
+                          xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+            configASSERT( false );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *  Run the workflow : trigger action on state and process receive path in MQTT loop
+ *
+ */
+static AzureIoTProvisioningClientResult_t prvAzureIoTProvisioningClientRunWorkflow( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle,
+                                                                                    uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTResult_t xResult;
+    AzureIoTProvisioningClientResult_t xReturn;
+    uint32_t ulWaitTime;
+
+    do
+    {
+        if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_COMPLETE )
+        {
+            break;
+        }
+
+        if( ulTimeoutMilliseconds > azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS )
+        {
+            ulTimeoutMilliseconds -= azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS;
+            ulWaitTime = azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS;
+        }
+        else
+        {
+            ulWaitTime = ulTimeoutMilliseconds;
+            ulTimeoutMilliseconds = 0;
+        }
+
+        prvTriggerAction( xAzureIoTProvisioningClientHandle );
+
+        if( ( xResult =
+                  AzureIoTMQTT_ProcessLoop( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ),
```

Comment:
That will break the contract in header [https://github.com/Azure/azure-iot-middleware-freertos/blob/f5e07fba53698f26ce03d172ce51134395ad8d59/source/interface/azure_iot_mqtt.h#L329](https://github.com/Azure/azure-iot-middleware-freertos/blob/f5e07fba53698f26ce03d172ce51134395ad8d59/source/interface/azure_iot_mqtt.h#L329).


Atleast in case of coreMQTT this will be a duplicate code at middleware layer.

---
In reply to: [603458639](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r603458639) [](ancestors = 603458639)

Reply:
OK - to keep simple agreed.  I'll resolve this conv then.